packages/supplementary-contracts/contracts/TokenUnlocking.sol

```diff
@@ -1,3 +1,332 @@
+// SPDX-License-Identifier: MIT
 pragma solidity 0.8.24;
 
-contract TokenUnlocking { }
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/utils/cryptography/ECDSA.sol";
+
+/// @title TokenUnlocking
+/// @notice Contract for managing Taiko token unlocking.
+///
+/// It manages only unlocking and vested tokens will be deposited into this contract (through
+/// 'depositToGrantee()' function) when the purchase notice sent out by Taiko, is paid. Unlocking
+/// will be a 4-year immutable period, with a 1-year cliff, counting from TGE.
+///
+/// Vesting will be a regular (quarterly / twice a year, TBD) 'off-chain' legal payment action,
+/// where those purschase notices can be exercised (and paid up-front, so before depoists made).
+/// Once tokens are deposited into this contract it cannot be forfeited anymore. If an
+/// employment is ended, the company (Taiko) simply does not send out purchase notices anymore, so
+/// that more tokens would not be deposited, beside the eligible proportion of that same vesting
+/// release period (e.g.: if Bob spent 1 month at Taiko out of that quarterly/half-yearly vesting,
+/// those will be deposited of course).
+///
+/// We should deploy multiple instances of this contract per grantee and per grant. So each person
+/// should have the same amount of contract deployed as grants granted (grant 1, grant 2, etc.)
+/// @custom:security-contact security@taiko.xyz
+contract TokenUnlocking is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+
+    // Variables, which shall be same for everybody, it is part of the GrantInfo struct.
+    struct UnlockTiming {
+        /// @notice Shall be TGE date.
+        uint64 unlockStartDate;
+        /// @notice The end date of cliff period. Shall be 1 year post TGE (date).
+        uint64 unlockCliffDate;
+        /// @notice Shall be 4 years (period)
+        uint32 unlockPeriod;
+    }
+
+    // Variables, which shall be unique per grantee
+    struct GrantInfo {
```

Comment:
Since each grantee will have his/her own deployment, we can also get rid of this struct and flatten it into the contract's storage layout directly.